presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -1209,6 +1257,24 @@ private boolean joinEqualityExpression(Expression expression, Collection<Symbol>
             return joinComparisonExpression(expression, leftSymbols, rightSymbols, ImmutableSet.of(EQUAL));
         }
 
+        private boolean joinDynamicFilteringExpression(Expression expression, Collection<Symbol> leftSymbols, Collection<Symbol> rightSymbols)
+        {
+            if (expression instanceof NotExpression) {
+                NotExpression notExpression = (NotExpression) expression;
+                boolean isDistinctFrom = joinComparisonExpression(notExpression.getValue(), leftSymbols, rightSymbols, ImmutableSet.of(IS_DISTINCT_FROM));
+                if (!isDistinctFrom) {
+                    return false;
+                }
+                ComparisonExpression comparison = (ComparisonExpression) notExpression.getValue();
+                Set<Type> expressionTypes = ImmutableSet.of(
+                        typeAnalyzer.getType(session, types, comparison.getLeft()),
+                        typeAnalyzer.getType(session, types, comparison.getRight()));
+                // Dynamic filtering is not supported with IS NOT DISTINCT FROM clause on REAL or DOUBLE types to avoid dealing with NaN values
```

Comment:
I think we might want to whitelist support for `IS NOT DISTINCT FROM` to types where we know that it can be translated into `EQUALS` with nulls. The reason is that within `io.prestosql.spi.function.OperatorType` there is a dedicated operator type for `EQUAL` and `IS_DISTINCT_FROM`, so these can have different implementations. For most of the types, we know the semantics, but for user types this might be different and we don't use `NOT(IS_DISTINCT_FROM)` explicitly within DF code.